File: chaos_analysis/lyapunov.m
% lyapunov.m - Chaos Analysis of 4D Discrete Sine-Cosine Hyperchaotic Map (4D-DSCHM)
% Computes and plots the Lyapunov Exponents spectrum and Bifurcation Diagram.

clear; clc; close all;

%% 1. Parameters Setup
b = 0.354; c = 1.287; d = 0.219; e = 1.543; f = 0.412; g = 1.109; h = 0.331;
a_start = 0.5;
a_end = 3.0;

%% 2. Run Lyapunov Exponents Spectrum Calculation
fprintf('Calculating Lyapunov Exponents Spectrum...\n');
le_steps = 150;
a_vals_le = linspace(a_start, a_end, le_steps);
iterations_le = 1500;
le_results = zeros(le_steps, 4);

for idx = 1:le_steps
    a = a_vals_le(idx);
    
    % Initial states
    x = 0.1; y = 0.2; z = 0.3; w = 0.4;
    Q = eye(4);
    sums = zeros(4, 1);
    
    % Transient warm-up
    for n = 1:500
        x = mod(sin(a * y) + b * cos(x) + w, 1.0);
        y = mod(sin(c * x) + d * cos(y), 1.0);
        z = mod(sin(e * z) + f * cos(w), 1.0);
        w = mod(sin(g * w) + h * cos(z), 1.0);
    end
    
    % Accumulation phase with QR decomposition
    for n = 1:iterations_le
        x = mod(sin(a * y) + b * cos(x) + w, 1.0);
        y = mod(sin(c * x) + d * cos(y), 1.0);
        z = mod(sin(e * z) + f * cos(w), 1.0);
        w = mod(sin(g * w) + h * cos(z), 1.0);
        
        J = get_jacobian(x, y, z, w, a, b, c, d, e, f, g, h);
        [Q, R] = qr(J * Q);
        
        diag_val = abs(diag(R));
        diag_val(diag_val < 1e-12) = 1e-12;
        sums = sums + log(diag_val);
    end
    
    le_results(idx, :) = (sums / iterations_le)';
end

%% 3. Run Bifurcation Diagram Calculation
fprintf('Calculating Bifurcation Diagram...\n');
bif_steps = 450;
a_vals_bif = linspace(a_start, a_end, bif_steps);
transient_bif = 800;
keep_bif = 150;

% Prepare arrays for fast plotting
a_out = zeros(bif_steps * keep_bif, 1);
x_out = zeros(bif_steps * keep_bif, 1);

counter = 1;
for idx = 1:bif_steps
    a = a_vals_bif(idx);
    x = 0.1; y = 0.2; z = 0.3; w = 0.4;
    
    % Transient discard
    for n = 1:transient_bif
        x = mod(sin(a * y) + b * cos(x) + w, 1.0);
        y = mod(sin(c * x) + d * cos(y), 1.0);
        z = mod(sin(e * z) + f * cos(w), 1.0);
        w = mod(sin(g * w) + h * cos(z), 1.0);
    end
    
    % Collect points
    for n = 1:keep_bif
        x = mod(sin(a * y) + b * cos(x) + w, 1.0);
        y = mod(sin(c * x) + d * cos(y), 1.0);
        z = mod(sin(e * z) + f * cos(w), 1.0);
        w = mod(sin(g * w) + h * cos(z), 1.0);
        
        a_out(counter) = a;
        x_out(counter) = x;
        counter = counter + 1;
    end
end

%% 4. Plotting Results
figure('Name', 'Chaos Analysis of 4D-DSCHM', 'Position', [100, 100, 1000, 800]);

% Subplot 1: Lyapunov Exponents
subplot(2, 1, 1);
plot(a_vals_le, le_results(:, 1), 'r-', 'LineWidth', 1.5, 'DisplayName', '\lambda_1 (LE1)');
hold on;
plot(a_vals_le, le_results(:, 2), 'b-', 'LineWidth', 1.5, 'DisplayName', '\lambda_2 (LE2)');
plot(a_vals_le, le_results(:, 3), 'g-', 'LineWidth', 1.5, 'DisplayName', '\lambda_3 (LE3)');
plot(a_vals_le, le_results(:, 4), 'm-', 'LineWidth', 1.5, 'DisplayName', '\lambda_4 (LE4)');
yline(0, 'k--', 'LineWidth', 1.2);
xlabel('Control Parameter \alpha');
ylabel('Lyapunov Exponents');
title('Lyapunov Exponents Spectrum of the 4D-DSCHM');
legend('Location', 'northeast');
grid on;
set(gca, 'FontSize', 11);

% Subplot 2: Bifurcation Diagram
subplot(2, 1, 2);
scatter(a_out, x_out, 1.5, [0, 0, 0.5], 'filled', 'MarkerFaceAlpha', 0.4);
xlabel('Control Parameter \alpha');
ylabel('State Variable x');
title('Bifurcation Diagram of the 4D-DSCHM');
xlim([a_start, a_end]);
ylim([0, 1.0]);
grid on;
set(gca, 'FontSize', 11);

sgtitle('4D Sine-Cosine Hyperchaotic Map Chaos Analysis', 'FontSize', 14, 'FontWeight', 'bold');

%% Helper Jacobian Function
function J = get_jacobian(x, y, z, w, a, b, c, d, e, f, g, h)
    J = zeros(4, 4);
    J(1, 1) = -b * sin(x);
    J(1, 2) = a * cos(a * y);
    J(1, 3) = 0.0;
    J(1, 4) = 1.0;
    
    J(2, 1) = c * cos(c * x);
    J(2, 2) = -d * sin(y);
    J(2, 3) = 0.0;
    J(2, 4) = 0.0;
    
    J(3, 1) = 0.0;
    J(3, 2) = 0.0;
    J(3, 3) = e * cos(e * z);
    J(3, 4) = -f * sin(w);
    
    J(4, 1) = 0.0;
    J(4, 2) = 0.0;
    J(4, 3) = -h * sin(z);
    J(4, 4) = g * cos(g * w);
end

File: chaos_analysis/lyapunov.py
import numpy as np
import matplotlib.pyplot as plt

def get_jacobian(x, y, z, w, a, b, c, d, e, f, g, h_param):
    """
    Computes the 4x4 Jacobian matrix for the 4D-DSCHM.
    """
    J = np.zeros((4, 4))
    
    # Row 1 partial derivatives
    J[0, 0] = -b * np.sin(x)
    J[0, 1] = a * np.cos(a * y)
    J[0, 2] = 0.0
    J[0, 3] = 1.0
    
    # Row 2 partial derivatives
    J[1, 0] = c * np.cos(c * x)
    J[1, 1] = -d * np.sin(y)
    J[1, 2] = 0.0
    J[1, 3] = 0.0
    
    # Row 3 partial derivatives
    J[2, 0] = 0.0
    J[2, 1] = 0.0
    J[2, 2] = e * np.cos(e * z)
    J[2, 3] = -f * np.sin(w)
    
    # Row 4 partial derivatives
    J[3, 0] = 0.0
    J[3, 1] = 0.0
    J[3, 2] = -h_param * np.sin(z)
    J[3, 3] = g * np.cos(g * w)
    
    return J

def calculate_lyapunov_exponents(iterations=5000, start_a=0.5, end_a=3.0, steps=100):
    """
    Computes the four Lyapunov Exponents for varying values of parameter alpha (A).
    """
    a_values = np.linspace(start_a, end_a, steps)
    
    # System parameters
    b, c, d, e, f, g, h_param = 0.354, 1.287, 0.219, 1.543, 0.412, 1.109, 0.331
    
    le_results = np.zeros((steps, 4))
    
    for idx, a in enumerate(a_values):
        # Initial states
        x, y, z, w = 0.1, 0.2, 0.3, 0.4
        
        # Identity matrix for orthogonalization
        Q = np.eye(4)
        sums = np.zeros(4)
        
        # Warm-up phase
        for _ in range(500):
            x = (np.sin(a * y) + b * np.cos(x) + w) % 1.0
            y = (np.sin(c * x) + d * np.cos(y)) % 1.0
            z = (np.sin(e * z) + f * np.cos(w)) % 1.0
            w = (np.sin(g * w) + h_param * np.cos(z)) % 1.0
            
        # Calculation phase
        for _ in range(iterations):
            # 1. Iterate Map
            x = (np.sin(a * y) + b * np.cos(x) + w) % 1.0
            y = (np.sin(c * x) + d * np.cos(y)) % 1.0
            z = (np.sin(e * z) + f * np.cos(w)) % 1.0
            w = (np.sin(g * w) + h_param * np.cos(z)) % 1.0
            
            # 2. Get Jacobian
            J = get_jacobian(x, y, z, w, a, b, c, d, e, f, g, h_param)
            
            # 3. QR Orthogonalization to prevent exponent explosion
            M = np.dot(J, Q)
            Q, R = np.linalg.qr(M)
            
            # 4. Accumulate eigenvalues log
            diag_val = np.abs(np.diag(R))
            # Safe log calculation to avoid divide by zero warnings
            diag_val[diag_val < 1e-12] = 1e-12
            sums += np.log(diag_val)
            
        le_results[idx, :] = sums / iterations
        
    return a_values, le_results

def plot_lyapunov_exponents(a_values, le_results, output_path="results/plots/lyapunov_exponents.png"):
    """
    Plots and saves the Lyapunov exponents curve.
    """
    plt.figure(figsize=(10, 6))
    plt.plot(a_values, le_results[:, 0], label=r"$\lambda_1$ (LE1)", color='red')
    plt.plot(a_values, le_results[:, 1], label=r"$\lambda_2$ (LE2)", color='blue')
    plt.plot(a_values, le_results[:, 2], label=r"$\lambda_3$ (LE3)", color='green')
    plt.plot(a_values, le_results[:, 3], label=r"$\lambda_4$ (LE4)", color='purple')
    
    plt.axhline(0, color='black', linestyle='--', linewidth=1)
    plt.xlabel(r"Control Parameter $\alpha$")
    plt.ylabel("Lyapunov Exponents")
    plt.title("Lyapunov Exponents Spectrum of 4D-DSCHM")
    plt.grid(True, linestyle=':', alpha=0.6)
    plt.legend()
    plt.tight_layout()
    plt.savefig(output_path, dpi=300)
    plt.close()

if __name__ == "__main__":
    import os
    import sys
    sys.path.append(os.path.abspath(os.path.join(os.path.dirname(__file__), '..')))
    
    print("====================================================")
    print("Computing Lyapunov Exponents for 4D-DSCHM...")
    print("====================================================")
    
    os.makedirs("results/plots", exist_ok=True)
    
    # Compute and plot
    a_vals, le_results = calculate_lyapunov_exponents(iterations=1500, steps=80)
    plot_lyapunov_exponents(a_vals, le_results)
    print("LE analysis complete. Plot saved to results/plots/lyapunov_exponents.png")
    
    # Standard alpha parameter check
    idx = np.argmin(np.abs(a_vals - 1.412))
    le_vals = le_results[idx]
    
    def color_val(val):
        if val > 0:
            return f"\033[92m{val:+.6f}\033[0m"  # Green
        else:
            return f"\033[91m{val:+.6f}\033[0m"  # Red

    print(f"\nLyapunov Exponents at alpha = 1.412:")
    print(f"  LE1: {color_val(le_vals[0])}")
    print(f"  LE2: {color_val(le_vals[1])}")
    print(f"  LE3: {color_val(le_vals[2])}")
    print(f"  LE4: {color_val(le_vals[3])}")
    
    pos_count = np.sum(le_vals > 0.0)
    print(f"\nNumber of Positive Lyapunov Exponents: {pos_count}")
    if pos_count >= 2:
        print("Status: \033[92mHyperchaos Confirmed (>= 2 positive exponents)\033[0m")
    else:
        print("Status: \033[91mNo Hyperchaos\033[0m")
    print("====================================================")
